apps/api/src/services/distributed-lock/index.ts

```diff
@@ -0,0 +1,135 @@
+import Redis from 'ioredis';
+import Redlock, { Lock, ResourceLockedError, Settings } from 'redlock';
+import { Logger } from '@nestjs/common';
+
+const LOG_CONTEXT = 'DistributedLock';
+
+const redisUrl = `${process.env.REDIS_HOST}:${process.env.REDIS_PORT}` || 'localhost:6379';
+
+interface ILockOptions {
+  resource: string;
+  ttl: number;
+}
+
+let distributedLock: Redlock;
+let lockCounter = 0;
+
+export const startup = (
+  settings: Settings = {
+    automaticExtensionThreshold: 500,
+    driftFactor: 0.01,
+    retryCount: 50,
+    retryDelay: 100,
+    retryJitter: 200,
+  }
+): void => {
+  if (distributedLock) {
+    return;
+  }
+
+  // TODO: Implement distributed nodes (at least 3 Redis instances)
+  const rawInstances = [redisUrl].filter((instance) => !!instance).map((url) => new Redis(url));
+
+  distributedLock = new Redlock(rawInstances, settings);
+  Logger.log('Redlock started', LOG_CONTEXT);
+
+  /**
+   * https://github.com/mike-marcacci/node-redlock/blob/dc7bcd923f70f66abc325d23ae618f7caf01ad75/src/index.ts#L192
+   *
+   * Because Redlock is designed for high availability, it does not care if
+   * a minority of redis instances/clusters fail at an operation.
+   *
+   * However, it can be helpful to monitor and log such cases. Redlock emits
+   * an "error" event whenever it encounters an error, even if the error is
+   * ignored in its normal operation.
+   *
+   * This function serves to prevent Node's default behavior of crashing
+   * when an "error" event is emitted in the absence of listeners.
+   */
+  distributedLock.on('error', (error) => {
+    // Ignore cases where a resource is explicitly marked as locked on a client.
+    if (error instanceof ResourceLockedError) {
+      // TODO: Hide or move to trace.
+      Logger.error(error, LOG_CONTEXT);
```

Comment:
This logs the times some call hits an used lock. We do not need to log them but right now while we see how well the implementation works, might be useful to understand the process. Move it to `debug` level could be a good idea.